src/awkward/_nplikes/virtual.py

```diff
@@ -66,51 +69,89 @@ def __init__(
         self._dtype = np.dtype(dtype)
         self._array: Sentinel | ArrayLike = UNMATERIALIZED
         self._generator = generator
+        self._shape_generator = shape_generator
 
     @property
     def dtype(self) -> DType:
         return self._dtype
 
     @property
     def shape(self) -> tuple[ShapeItem, ...]:
+        self.get_shape()
         return self._shape
 
+    @property
+    def inner_shape(self) -> tuple[ShapeItem, ...]:
+        if len(self._shape) > 1:
+            return self.shape[1:]
+        return self._shape[1:]
+
     @property
     def ndim(self) -> int:
         return len(self._shape)
 
     @property
     def size(self) -> ShapeItem:
         size: ShapeItem = 1
-        for item in self._shape:
+        for item in self.shape:
             size *= item
         return size
 
     @property
     def nbytes(self) -> ShapeItem:
-        return self.size * self._dtype.itemsize
+        size: ShapeItem = 1
+        for item in self._shape:
+            size *= item
+        return size * self._dtype.itemsize
 
     @property
     def strides(self) -> tuple[ShapeItem, ...]:
         return self.materialize().strides  # type: ignore[attr-defined]
 
+    def get_shape(self) -> None:
+        if any(dim is unknown_length for dim in self._shape):
+            if self._shape_generator is not None:
+                shape = self._shape_generator()
+            else:
+                shape = self.materialize().shape
+            if len(shape) != len(self._shape):
+                raise TypeError(
+                    f"{type(self).__name__} had shape {self._shape} before materialization while the materialized array has shape {shape}"
+                )
+            for expected_dim, actual_dim in zip(self._shape, shape):
+                if expected_dim is not unknown_length and expected_dim != actual_dim:
+                    raise TypeError(
```

Comment:
I think this should be a `ValurError` instead

Reply:
and maybe check that all elements of `shape` are integers, i.e. `if not all(isinteger(dim) for dim in shape): raise ValueError(...)` ? (then the following `tuple(map(int, shape))` will also always work)